auditbeat/module/audit/file/monitor/recursive.go

```diff
@@ -0,0 +1,129 @@
+package monitor
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+)
+
+type recursiveWatcher struct {
+	inner  *fsnotify.Watcher
+	tree   FileTree
+	eventC chan fsnotify.Event
+	done   chan bool
+}
+
+func newRecursiveWatcher(inner *fsnotify.Watcher) *recursiveWatcher {
+	return &recursiveWatcher{
+		inner:  inner,
+		tree:   FileTree{},
+		eventC: make(chan fsnotify.Event, 1),
+	}
+}
+
+func (watcher *recursiveWatcher) Start() error {
+	watcher.done = make(chan bool, 1)
+	go watcher.forwardEvents()
+	return nil
+}
+
+func (watcher *recursiveWatcher) Add(path string) error {
+	return watcher.addRecursive(path)
+}
+
+func (watcher *recursiveWatcher) Close() error {
+	if watcher.done != nil {
+		// has been Started(), goroutine takes care of cleanup
+		watcher.done <- true
+		return nil
+	}
+	// not started
+	return watcher.close()
+}
+
+func (watcher *recursiveWatcher) EventChannel() <-chan fsnotify.Event {
+	return watcher.eventC
+}
+
+func (watcher *recursiveWatcher) ErrorChannel() <-chan error {
+	return watcher.inner.Errors
+}
+
+func (watcher *recursiveWatcher) addRecursive(path string) error {
+	return filepath.Walk(path, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return errors.Wrap(err, "file tree recursion failed")
+		}
+		if info.IsDir() {
+			// TODO: can be a symlink?
```

Comment:
You could add a test case for this to find out.

Reply:
done